app/src/main/java/ayds/lisboa/songinfo/moredetails/fulllogic/OtherInfoWindow.kt

```diff
@@ -0,0 +1,204 @@
+import com.google.gson.Gson
+import com.google.gson.JsonObject
+import android.content.Intent
+import android.net.Uri
+import android.os.Build.*
+import com.squareup.picasso.Picasso
+import android.text.Html
+import android.text.Spanned
+import android.util.Log
+import android.view.View
+import android.widget.ImageView
+import retrofit2.Response
+import java.io.IOException
+import java.lang.StringBuilder
+import android.widget.Button
+
+const val ARTIST_NAME_EXTRA = "artistName"
+private const val LASTFM_API = "https://ws.audioscrobbler.com/2.0/"
+private const val IMAGE_URL_LASTFM = "https://upload.wikimedia.org/wikipedia/commons/thumb/d/d4/Lastfm_logo.svg/320px-Lastfm_logo.svg.png"
+private const val ARTIST = "artist"
+private const val ARTIST_BIOGRAPHY = "bio"
+private const val ARTIST_BIOGRAPHY_EXTRACT = "content"
+private const val ARTIST_BIOGRAPHY_URL = "url"
+private const val ASTERISK = "[*]"
+private const val NO_RESULTS = "No Results"
+
+class OtherInfoWindow : AppCompatActivity() {
+
+    private lateinit var biographyTextView: TextView
+    private lateinit var openUrlButton: Button
+    private lateinit var imageView: ImageView
+    private lateinit var dataBase: DataBase
+    private lateinit var biographyJsonObject: JsonObject
+    private lateinit var serviceApi: LastFMAPI
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_other_info)
+
+        initProperties()
+        initDataBase()
+        initArtist()
+    }
+
+    private fun initProperties() {
+        biographyTextView = findViewById(R.id.biographyTextView)
+        openUrlButton = findViewById<View>(R.id.openUrlButton) as Button
+        imageView = findViewById<View>(R.id.imageView) as ImageView
+    }
+
+    private fun initDataBase(){
+        dataBase = DataBase(this)
+        setLastFMAPI()
+    }
+
+    private fun setLastFMAPI() {
+        serviceApi = initRetrofit().create(LastFMAPI::class.java)
+    }
+
+    private fun initRetrofit() = Retrofit.Builder()
+        .baseUrl(LASTFM_API)
+        .addConverterFactory(ScalarsConverterFactory.create())
+        .build()
+
+    private fun initArtist() {
+        val artist = getArtist()
+        getArtistInfo(artist)
+    }
+
+    private fun getArtist(): String? {
+        return intent.getStringExtra(ARTIST_NAME_EXTRA)
+    }
+
+    private fun getArtistInfo(artistName: String?) {
+        Thread {
+            artistInfo(artistName)
+        }.start()
+    }
+
+    private fun artistInfo(artistName: String?) {
+        val artistBiography = createBiography(artistName)
+        updateArtistImage()
+        updateArtistBiography(artistBiography)
+    }
+
+    private fun createBiography(artistName: String?): String {
+        val biographyText = dataBase.getInfo(artistName)
+        return if (biographyText != null) "$ASTERISK $biographyText" else getArtistBiographyFromLastFM(artistName)
```

Comment:
Este método debería contener el algoritmo del patrón repository. Lo que les falta en sí es guardar el resultado en la db, cosa que les quedó como side effect varios niveles más abajo.
También mejorar el nombre.

Reply:
En esta funcion, deberiamos llamar a la funcion saveDataBase? La duda es como modificarlo, ya que esta en este orden porque es necesario pasarlo primero a HTLM, por eso el saveDataBase esta en la funcion updateArtistBiographyText.

Reply:
La idea es extraer esa lógica y llegar al algoritmo de repository. Algo así debería quedar:
```suggestion
    private fun createBiography(artistName: String?): String {
        var biographyText = dataBase.getInfo(artistName)
        if (biographyText != null) 
           biographyText = "$ASTERISK $biographyText" 
        else {
           biographyText = getArtistBiographyFromLastFM(artistName)
           saveDataBase(artistName, biographyText)
        }
        
        return biographyText
```